src/Nethermind/Nethermind.Evm/EvmStack.cs

```diff
@@ -13,165 +13,178 @@
 using System.Diagnostics.CodeAnalysis;
 using System.Runtime.Intrinsics;
 using System.Diagnostics;
+using System.Runtime.Intrinsics.X86;
 
 namespace Nethermind.Evm
 {
+    using static Nethermind.Evm.VirtualMachine;
+
     using Word = System.Runtime.Intrinsics.Vector256<byte>;
 
-    public ref struct EvmStack
+    public ref struct EvmStack<TTracing>
+        where TTracing : struct, IIsTracing
     {
         public const int RegisterLength = 1;
         public const int MaxStackSize = 1025;
         public const int ReturnStackSize = 1023;
+        public const int WordSize = 32;
+        public const int AddressSize = 20;
 
         public EvmStack(scoped in Span<byte> bytes, scoped in int head, ITxTracer txTracer)
         {
             _bytes = bytes;
             Head = head;
             _tracer = txTracer;
-            Register = _bytes.Slice(MaxStackSize * 32, 32);
-            Register.Clear();
         }
 
         public int Head;
 
-        public Span<byte> Register;
-
         private Span<byte> _bytes;
 
         private ITxTracer _tracer;
 
         public void PushBytes(scoped in Span<byte> value)
         {
-            if (_tracer.IsTracingInstructions) _tracer.ReportStackPush(value);
+            if (typeof(TTracing) == typeof(IsTracing)) _tracer.ReportStackPush(value);
```

Comment:
Why? Seems counter productive.

Reply:
The generic is a struct parameter (rather than class) so the Jit generates a method for each type; the type checks are evaluated at Jit time completely eliminating the if and and code behind it (if the if doesn't pass)

So if the TTracing isn't IsTracing then all the code for tracing is completely removed from the asm.

Note: this _only_ works for struct types; classes and interfaces use shared generics and only have one set of asm generated for all types.

Reply:
Ah nevermind.

Reply:
But can use generics to do compile time branch elimination and have an instrumented method and fast non-instrumented method just by switching the generic parameter